src/database_client.py

```diff
@@ -0,0 +1,117 @@
+from psycopg2 import sql
+from psycopg2.extensions import connection
+
+from charm_types import DatasourcePostgreSQL
+from exceptions import CharmDatabaseRelationNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseClient:
+    """A class representing the Synapse application."""
+
+    def __init__(
+        self, datasource: typing.Optional[DatasourcePostgreSQL], alternative_database: str = ""
+    ):
+        """Initialize a new instance of the Synapse class.
+
+        Args:
+            datasource: datasource to use to connect.
+            alternative_database: database to connect to.
+                The default is to use the one provided by datasource.
+
+        Raises:
+            CharmDatabaseRelationNotFoundError: if there is no relation.
+        """
+        if datasource is None:
+            raise CharmDatabaseRelationNotFoundError("No database relation was found.")
+        self._datasource = datasource
+        self._database_name = datasource["db"]
+        self._alternative_database = alternative_database
+        self._conn: connection = None
+
+    def _connect(self) -> None:
+        """Get connection.
+
+        Raises:
+            Error: something went wrong while connecting to the database.
+        """
+        if self._conn is None or self._conn.closed != 0:
+            logger.debug("Connecting to database")
+            try:
+                user = self._datasource["user"]
+                password = self._datasource["password"]
+                host = self._datasource["host"]
+                database_name = (
+                    self._alternative_database
+                    if self._alternative_database
+                    else self._datasource["db"]
+                )
+                self._conn = psycopg2.connect(
+                    f"dbname='{database_name}' user='{user}' host='{host}'"
+                    f" password='{password}' connect_timeout=5"
+                )
+                self._conn.autocommit = True
+            except psycopg2.Error as exc:
+                logger.exception("Failed to connect to database: %s", str(exc))
+                raise
+
+    def _close(self) -> None:
+        """Close database connection."""
+        if self._conn is not None:
+            self._conn.close()
+            self._conn = None
+
+    def prepare(self) -> None:
+        """Change database collate and ctype as required by Synapse.
+
+        Raises:
+            Error: something went wrong while preparing the database.
+        """
+        try:
+            self._connect()
+            with self._conn.cursor() as curs:
+                curs.execute(
+                    sql.SQL(
+                        "UPDATE pg_database SET datcollate='C', datctype='C' WHERE datname = {}"
+                    ).format(sql.Literal(self._database_name))
+                )
+        except psycopg2.Error as exc:
+            logger.exception("Failed to prepare database: %r", exc)
+            raise
```

Comment:
Based on this discussion, it could be even removed. It was left only for logging reason:
https://github.com/canonical/synapse-operator/pull/13#discussion_r1253285244

I agree regarding the library but Synapse only supports sqlite3 (for SQLite, local file) or psycopg2 (for PostgreSQL).

Reply:
Related: #52